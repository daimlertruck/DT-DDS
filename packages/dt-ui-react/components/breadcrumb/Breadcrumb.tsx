import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { ArrowRightIcon } from '../../core';
import { BaseProps } from '../../types';
import { Tooltip } from '../tooltip';

import {
  BreadcrumbStyled,
  ListStyled,
  ItemStyled,
  ItemContainerStyled,
} from './Breadcrumb.styled';
import { useCollapsedBreadcrumb } from './hooks/useCollapsedBreadcrumb';

const BREADCRUMB_CHARACTER_LIMIT = 30;

type BreadcrumbSeparator = 'slash' | 'arrow';

export interface BreadcrumbProps extends BaseProps {
  separator?: BreadcrumbSeparator;
}

export const Breadcrumb = ({
  children,
  separator = 'slash',
}: BreadcrumbProps) => {
  const { visibleChildren, breadcrumbRef, breadcrumbListRef } =
    useCollapsedBreadcrumb(children);

  return (
    <BreadcrumbStyled
      ref={breadcrumbRef}
      aria-label='Breadcrumb'
      data-testid='breadcrumb'
    >
      <ListStyled ref={breadcrumbListRef}>
        {withSeparator(visibleChildren, separator)}
      </ListStyled>
    </BreadcrumbStyled>
  );
};

const BreadcrumbItem = ({ children, style }: BaseProps) => {
  const { truncatedChildren, tooltip } = mapTruncateChildren(children);

  return (
    <ItemStyled style={style}>
      <Tooltip style={{ overflow: 'hidden' }}>
        <ItemContainerStyled>{truncatedChildren}</ItemContainerStyled>
        <Tooltip.Content>{tooltip}</Tooltip.Content>
      </Tooltip>
    </ItemStyled>
  );
};

const truncateText = (text: string) => {
  if (text.length >= BREADCRUMB_CHARACTER_LIMIT) {
    return `${text.substring(0, BREADCRUMB_CHARACTER_LIMIT)}...`;
  }
  return text;
};

const mapTruncateChildren = (
  children: ReactNode
): { tooltip: string; truncatedChildren: ReactNode } => {
  let tooltip = '';
  return {
    truncatedChildren: Children.map(children, (child) => {
      if (typeof child === 'string') {
        const truncatedText = truncateText(child);
        tooltip = truncatedText !== child ? child : '';

        return truncatedText;
      } else if (isValidElement(child)) {
        const { truncatedChildren, tooltip: childTooltip } =
          mapTruncateChildren(child.props.children);
        tooltip = childTooltip;

        return cloneElement(child, {
          ...child.props,
          children: truncatedChildren,
        });
      } else {
        return null;
      }
    }),
    tooltip,
  };
};

const withSeparator = (children: ReactNode, separator: BreadcrumbSeparator) => {
  const addSeparator = () =>
    ({
      slash: <span>/</span>,
      arrow: <ArrowRightIcon />,
    }[separator]);

  const isLastBreadcrumb = (index: number) =>
    Children.count(children) - 1 === index;

  return Children.map(children, (child, index) => (
    <>
      {child}
      {!isLastBreadcrumb(index) && addSeparator()}
    </>
  ));
};

Breadcrumb.Item = BreadcrumbItem;
