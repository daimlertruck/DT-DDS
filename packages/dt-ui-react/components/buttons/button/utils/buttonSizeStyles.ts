import { CustomTheme as Theme } from '../../../../types';
import { ButtonSizes } from '../constants/types';

export const buttonSizeStyles = (size: ButtonSizes, theme: Theme) =>
  ({
    large: `
      ${theme.fontStyles.button1}
      padding: ${theme.spacing['2xs']} ${theme.spacing.s};
    `,
    medium: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['3xs']} ${theme.spacing.xs};
    `,
    small: `
      ${theme.fontStyles.button2}
      padding: ${theme.spacing['4xs']} ${theme.spacing['2xs']};
    `,
  }[size]);
