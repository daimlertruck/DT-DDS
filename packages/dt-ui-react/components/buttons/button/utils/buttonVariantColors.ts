import { CustomTheme as Theme } from '../../../../types';
import {
  ButtonVariant,
  ContainedColors,
  OutlinedColors,
  TextColors,
} from '../constants/types';

interface ColorKeys {
  color: string;
  bgColor: string;
  hoverBgColor?: string;
  hoverColor?: string;
}

export const buttonVariantColors = (
  color: ContainedColors | OutlinedColors,
  theme: Theme
): Record<ButtonVariant, ColorKeys> => ({
  contained: {
    color: theme.colors.alpha.white,
    bgColor: theme.palette.alpha[color],
    hoverBgColor: theme.palette.alpha[`${color}Dark_300` as ContainedColors],
  },
  outlined: {
    color:
      color === 'neutral'
        ? theme.colors.alpha.gray_200
        : theme.palette.alpha[color],
    bgColor: theme.colors.alpha.white,
    hoverBgColor: theme.palette.alpha[`${color}Light_300` as OutlinedColors],
  },
  text: {
    color: theme.palette.alpha[color],
    bgColor: 'transparent',
    hoverColor: theme.palette.alpha[`${color as TextColors}Dark_300`],
  },
});
