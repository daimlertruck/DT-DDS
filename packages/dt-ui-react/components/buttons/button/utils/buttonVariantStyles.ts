import { CustomTheme as Theme } from '../../../../types';
import { ButtonVariant, OutlinedColors, SolidColors } from '../constants/types';

export const buttonVariantStyles = (
  variant: ButtonVariant,
  theme: Theme,
  color: SolidColors | OutlinedColors
) =>
  ({
    solid: `
      text-transform: uppercase;
      border: ${theme.spacing.none};

      &:focus {
        box-shadow: 0px 2px 1px 0px #1A1A1A0D, 0px 0px 0px 4px ${theme.colors.blue_10};
        background-color: ${theme.palette[color]};
      }

      &:disabled {
        opacity: 38%;
      }
  `,
    outlined: `
      border: 1px solid;
      text-transform: uppercase;
        ${
          color === 'neutral'
            ? theme.palette.neutralDark_100
            : theme.palette[color]
        };

      &:disabled {
        color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        background-color: ${theme.colors.white};
        border-color: ${
          color === 'neutral'
            ? theme.palette.neutralLight_100
            : theme.palette[color]
        };
        cursor: default;
      }
    `,
    text: `
      border: none;

      &:disabled {
        color: ${theme.palette.neutralLight_100};
      }
    `,
  }[variant]);
