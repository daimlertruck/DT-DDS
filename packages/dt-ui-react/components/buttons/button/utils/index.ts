export * from './buttonVariantColors';
export * from './buttonVariantStyles';
