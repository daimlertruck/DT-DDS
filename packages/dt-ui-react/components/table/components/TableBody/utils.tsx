import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { TableDataCellHeader } from '../TableDataCell/TableDataCell';

export const addHeadersToDataCells = (
  dataCells: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(dataCells, (dataCell, index) => {
    if (isValidElement(dataCell)) {
      const header = headers[index];
      return cloneElement(
        dataCell,
        {},
        <>
          <TableDataCellHeader>{header}</TableDataCellHeader>
          {dataCell.props.children}
        </>
      );
    }
    return dataCell;
  });
};

export const addHeadersToDataCellsInRows = (
  rows: ReactNode,
  headers: ReactNode[]
) => {
  return Children.map(rows, (row) => {
    if (isValidElement(row)) {
      return cloneElement(
        row,
        {},
        addHeadersToDataCells(row.props.children, headers)
      );
    }
    return row;
  });
};
