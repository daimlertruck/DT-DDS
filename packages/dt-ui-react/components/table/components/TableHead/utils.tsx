import { Children, ReactNode, isValidElement } from 'react';

const extractColumnHeaders = (columnHeaders: ReactNode): ReactNode[] => {
  return Children.toArray(columnHeaders).reduce<ReactNode[]>(
    (headers, columnHeader) => {
      if (isValidElement(columnHeader)) {
        return [...headers, columnHeader.props.children];
      }

      return headers;
    },
    []
  );
};

export const extractTableHeaders = (theadRow: ReactNode) => {
  if (isValidElement(theadRow)) {
    return extractColumnHeaders(theadRow.props.children);
  }

  return [];
};
