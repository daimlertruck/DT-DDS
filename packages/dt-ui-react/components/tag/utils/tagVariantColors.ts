import { CustomTheme as Theme } from '../../../types';
import { TagColor, TagVariant } from '../constants';

interface TagVariantColors {
  border: string | number;
  color: string;
  borderColor?: string;
  backgroundColor: string;
  '&:hover'?: {
    borderColor?: string;
    backgroundColor: string;
    color: string;
  };
}

interface TagVariantColorsProps {
  theme: Theme;
  color: TagColor;
  variant: TagVariant;
  hasHover: boolean;
  isDisabled: boolean;
}

const getColorWithShade = ({
  theme,
  color,
  shade,
}: Pick<TagVariantColorsProps, 'color'> & {
  shade: '90' | '60' | '30' | '10';
  theme: Theme;
}) => {
  const { colors, palette } = theme;

  if (color === 'primary') {
    return {
      '90': palette.primaryDark_500,
      '60': palette.primaryDark_200,
      '30': palette.primaryLight_100,
      '10': palette.primaryLight_400,
    }[shade];
  }

  const colorWithShade: keyof Theme['colors'] = `${color}_${shade}`;

  return colors[colorWithShade];
};

export const tagVariantColors = ({
  theme,
  color,
  variant,
  hasHover,
  isDisabled,
}: TagVariantColorsProps): TagVariantColors => {
  const colorShade90 = getColorWithShade({
    theme,
    color,
    shade: '90',
  });
  const colorShade60 = getColorWithShade({
    theme,
    color,
    shade: '60',
  });
  const colorShade30 = getColorWithShade({
    theme,
    color,
    shade: '30',
  });
  const colorShade10 = getColorWithShade({
    theme,
    color,
    shade: '10',
  });

  const colorWithDisabled = isDisabled ? colorShade30 : colorShade60;

  return {
    solid: {
      border: theme.spacing.none,
      color: theme.colors.white,
      backgroundColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          backgroundColor: colorShade90,
          color: theme.colors.white,
        },
      }),
    },
    outlined: {
      border: `1px solid ${colorShade60}`,
      color: colorWithDisabled,
      backgroundColor: colorShade10,
      borderColor: colorWithDisabled,
      ...(hasHover && {
        '&:hover': {
          borderColor: colorShade90,
          backgroundColor: colorShade10,
          color: colorShade90,
        },
      }),
    },
  }[variant];
};
