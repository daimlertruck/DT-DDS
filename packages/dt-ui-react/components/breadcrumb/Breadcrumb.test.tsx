import { render } from '@testing-library/react';
import React from 'react';

import { ViewAgendaIcon } from '../../core';
import { withProviders } from '../../utils';

import { Breadcrumb } from './Breadcrumb';

describe('<Breadcrumb /> component', () => {
  const ProvidedBreadcrumb = withProviders(Breadcrumb);

  beforeEach(() => {
    jest.mock('../../hooks/useMedia', () => ({
      useMedia: jest.fn(() => true),
    }));
  });

  it('renders nav element with the correct breadcrumb items', () => {
    const breadcrumbs = [
      { name: 'Name 1', showIcon: true },
      {
        name: 'Second page has a long text to describe its path',
      },
      { disabled: true, name: 'Name 3' },
      {
        name: 'Last page has a long text to describe its path',
      },
    ];

    const { container } = render(
      <ProvidedBreadcrumb separator={'arrow'}>
        {breadcrumbs.map((breadcrumb, i) => (
          <Breadcrumb.Item
            key={`${i}-${breadcrumb.name}`}
            disabled={breadcrumb.disabled}
          >
            {breadcrumb.showIcon && <ViewAgendaIcon />}
            <a href='/'>{breadcrumb.name}</a>
          </Breadcrumb.Item>
        ))}
      </ProvidedBreadcrumb>
    );

    expect(container).toMatchSnapshot();
  });
});
