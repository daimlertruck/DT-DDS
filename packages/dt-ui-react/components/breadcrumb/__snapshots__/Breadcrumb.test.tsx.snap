// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumb /> component renders nav element with the correct breadcrumb items 1`] = `
.emotion-0 {
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.34;
  margin: 0;
  color: #8B8C8C;
}

.emotion-2 {
  padding: 0px;
  margin: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  list-style: none;
}

.emotion-4:last-child {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.emotion-4:not(:last-child):after {
  content: '/';
  padding: 8px;
}

.emotion-6 {
  color: #0078D6;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <nav
    aria-label="Breadcrumb"
    class="emotion-0 emotion-1"
    data-testid="breadcrumb"
  >
    <ul
      class="emotion-2 emotion-3"
    >
      <li
        class="emotion-4 emotion-5"
      >
        <a
          class="emotion-6 emotion-7"
          href="/"
        >
          Name 1
        </a>
      </li>
      <li
        class="emotion-4 emotion-5"
      >
        Name 2
      </li>
    </ul>
  </nav>
</div>
`;
